src/plots/cartesian/align_period.js

```diff
@@ -0,0 +1,113 @@
+/**
+* Copyright 2012-2020, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var isNumeric = require('fast-isnumeric');
+var Lib = require('../../lib');
+var ms2DateTime = Lib.ms2DateTime;
+var dateTime2ms = Lib.dateTime2ms;
+var constants = require('../../constants/numerical');
+var ONEDAY = constants.ONEDAY;
+var ONEAVGMONTH = constants.ONEAVGMONTH;
+var ONEAVGYEAR = constants.ONEAVGYEAR;
+
+module.exports = function alignPeriod(trace, ax, axLetter, vals) {
+    if(ax.type !== 'date') return vals;
+
+    var alignment = trace[axLetter + 'periodalignment'];
+    if(!alignment) return vals;
+
+    var period = trace[axLetter + 'period'];
+    if(isNumeric(period)) {
+        period = +period;
+        if(period <= 0) return vals;
+    } else if(typeof period === 'string' && period.charAt(0) === 'M') {
+        var n = +(period.substring(1));
+        if(n > 0 && Math.round(n) === n) period = n * ONEAVGMONTH;
+        else return vals;
+    }
+
+    if(period > 0) {
+        var calendar = ax.calendar;
+
+        var isStart = 'start' === alignment;
+        // var isMiddle = 'middle' === alignment;
+        var isEnd = 'end' === alignment;
+
+        var offset = (new Date()).getTimezoneOffset() * 60000;
+        var period0 = trace[axLetter + 'period0'];
+        var base = (dateTime2ms(period0, calendar) || 0) - offset;
+
+        var newVals = [];
+        var len = vals.length;
+        for(var i = 0; i < len; i++) {
+            var v = vals[i] - base;
+
+            var dateStr = ms2DateTime(v, 0, calendar);
+            var O = new Date(dateStr);
+            var year = O.getFullYear();
+            var month = O.getMonth();
+            var day = O.getDate();
+
+            var newD;
+            var startTime;
+            var endTime;
+
+            var nMonths = Math.floor(period / ONEAVGMONTH) % 12;
+            var nYears = Math.floor((period - nMonths * ONEAVGMONTH) / ONEAVGYEAR);
+            var nDays = Math.floor((period - nMonths * ONEAVGMONTH - nYears * ONEAVGYEAR) / ONEDAY);
```

Comment:
I think this approach using `ONEAVGMONTH` is going to cause problems at some point. I would not mix together the numeric periods and month-based periods, I'd make completely separate code for each case. That way if someone really wants to have 90-day periods or something, they can do that by specifying the `period` in milliseconds, and we won't accidentally convert it to months.

The logic should be extremely simple when `period` is specified in milliseconds, and should not need to involve any date functions at all once `period0` is also converted to milliseconds.

When period is specified in months, I'd encourage you to also reuse existing functionality, specifically [`Lib.incrementMonth`](https://github.com/plotly/plotly.js/blob/master/src/lib/dates.js#L497) because that will ensure consistent behavior around various edge cases like `period0` being set at the end of the month. Not that we care particularly, but it'll probably also give you world calendar support for free. For a given `v` you should be able to estimate the number of periods away from `period0` it is, then use `Lib.incrementMonth(period0, nEstimated*periodMonths, calendar)` to get close, and finally iterate forward or backward until you have the closest period boundary on either side of the datum => those are `startTime` and `endTime`.

Reply:
Good call. Done in https://github.com/plotly/plotly.js/pull/5074/commits/afa896514e7dbd0defebc2823c9916e76a710ad5.